File: 01-lambda-s3-events/lambda_workspace_python/lambda_script_sha256.py
import hashlib
import boto3

# calculate the SHA-256 hash
def calculate_sha256(file_path):
    with open(file_path, 'rb') as f:
        sha256_hasher = hashlib.sha256()
        for chunk in iter(lambda: f.read(4096), b''):
            sha256_hasher.update(chunk)
        return sha256_hasher.hexdigest()

# upload the SHA-256 hash to S3 bucket
def upload_sha256_hash(bucket, key, sha256_hash):
    s3_client = boto3.client('s3')
    s3_client.put_object(Bucket=bucket, Key=key, Body=sha256_hash.encode('utf-8'))

# lamda handler processing triggered by subscribed S3 event
def lambda_handler(event, context):
    # explicitly receive SNS object from records, extract message
    # and parse JSON into Python dictionary
    #s3_event = json.loads(event['Records'][0]['Sns']['Message'])

    # parse JSON into Python dictionary
    #s3_event = json.loads(event["Records"][0]["body"])

    # extract S3 bucket name and key from the event
    file_name = event['Records'][0]['s3']['object']['key']
    bucket_name = event['Records'][0]['s3']['bucket']['name']

    # perform SHA-256 calculation for file stored on S3 bucket
    # for which S3 creation event has been triggered
    sha256_hash = calculate_sha256(f's3://{bucket_name}/{file_name}')

    # store calculated SHA-256 hash as file on same S3 bucket
    # for which S3 creation event has been triggered
    upload_sha256_hash(bucket_name, f'{file_name}.sha256', sha256_hash)

File: 01-lambda-s3-events/lambda_workspace_java/lambda_script_sha256.java
import java.io.BufferedInputStream;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.security.MessageDigest;

import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.model.PutObjectRequest;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.amazonaws.services.lambda.runtime.events.S3Event;

public class Lambda implements RequestHandler<S3Event, String> {

    private static final int BUFFER_SIZE = 4096;

    public static String calculateSha256(String filePath) {
        MessageDigest messageDigest = null;
        try {
            messageDigest = MessageDigest.getInstance("SHA-256");
        } catch (Exception e) {
            throw new RuntimeException("Failed to initialize SHA-256 message digest", e);
        }

        try (InputStream inputStream = new FileInputStream(filePath)) {
            byte[] data = new byte[BUFFER_SIZE];
            int read;
            while ((read = inputStream.read(data)) != -1) {
                messageDigest.update(data, 0, read);
            }
        } catch (IOException e) {
            throw new RuntimeException("Failed to read file for SHA-256 calculation", e);
        }

        byte[] sha256HashBytes = messageDigest.digest();
        return new String(sha256HashBytes, "UTF-8");
    }

    public static void uploadSha256hash(String bucketName, String key, String sha256Hash) {
        AmazonS3 s3Client = new AmazonS3Client();
        PutObjectRequest putObjectRequest = new PutObjectRequest(bucketName, key, sha256Hash.getBytes("UTF-8"));
        s3Client.putObject(putObjectRequest);
    }

    @Override
    public String lambda_handler(S3Event s3Event, Context context) {
        S3EventNotification notification = s3Event.getRecords().get(0);
        String bucketName = notification.getS3().getBucket().getName();
        String fileName = notification.getS3().getObject().getKey();

        try {
            // Calculate the SHA-256 hash of the file
            String sha256Hash = calculateSha256("s3://" + bucketName + "/" + fileName);

            // Store the calculated SHA-256 hash as a file in the same S3 bucket
            uploadSha256hash(bucketName, fileName + ".sha256", sha256Hash);
        } catch (Exception e) {
            e.printStackTrace();
            return "Failed to process S3 objectCreated event: " + e.getMessage();
        }
        return "Successfully processed S3 objectCreated event.";
    }
}